### ISSUE TITLE
[halo-pr-prep] Task 8: Create end-to-end integration test

### ISSUE THREAD

**Issue by jalengg** [jalengg/PyHealth#15]
## Task 8: Create end-to-end integration test

**Depends on:** Tasks 2 & 3 (task function and refactored model)

### Goal

Create comprehensive integration tests that verify the complete HALO workflow: load data, train, generate.

### Files to Create

- **Create:** \`tests/integration/test_halo_end_to_end.py\`

#### Step 1: Create integration test file

Create \`tests/integration/test_halo_end_to_end.py\` with 2 tests:

**Test 1: \`test_halo_end_to_end_workflow\`**
- Marked with \`@pytest.mark.integration\`
- Skipped if MIMIC-III data not available
- Tests: load dataset → set task → split data → train model (1 epoch) → generate synthetic data
- Verifies checkpoint saved and synthetic output format

**Test 2: \`test_halo_processors_compatibility\`**
- Tests HALO works with NestedSequenceProcessor
- Uses mock samples (no MIMIC-III required)
- Verifies processors created correctly
- Tests forward pass with batched data

#### Step 2: Run integration tests

\`\`\`bash
pytest tests/integration/test_halo_end_to_end.py -v -m integration
\`\`\`
Expected: PASS (if MIMIC-III available) or SKIP (if not available)

#### Step 3: Commit

\`\`\`bash
git add tests/integration/test_halo_end_to_end.py
git commit -m \"test: add end-to-end integration tests for HALO\"
\`\`\`

### Test Structure

\`\`\`python
@pytest.mark.integration
@pytest.mark.skipif(
    not os.path.exists(\"/srv/local/data/physionet.org/files/mimiciii/1.4\"),
    reason=\"MIMIC-III data not available\"
)
def test_halo_end_to_end_workflow():
    # Load dataset
    base_dataset = MIMIC3Dataset(...)
    sample_dataset = base_dataset.set_task(halo_generation_mimic3_fn)
    
    # Split data
    train_dataset, val_dataset, _ = split_by_patient(...)
    
    # Train model (1 epoch, small config)
    model = HALO(dataset=sample_dataset, embed_dim=64, n_heads=2, n_layers=1, epochs=1)
    model.train_model(train_dataset, val_dataset)
    
    # Generate synthetic data
    synthetic_patients = model.synthesize_dataset(num_samples=10)
    
    # Verify output format
    assert len(synthetic_patients) == 10
    assert \"visits\" in synthetic_patients[0]
    assert len(synthetic_patients[0][\"visits\"]) > 0
\`\`\`

### Success Criteria

- [ ] Integration tests created
- [ ] Tests marked with \`@pytest.mark.integration\`
- [ ] Skip if MIMIC-III data unavailable
- [ ] End-to-end workflow test passes
- [ ] Processor compatibility test passes
- [ ] Synthetic data format validated

### Related

- Plan: \`docs/plans/2026-02-16-halo-pyhealth-integration.md\` (lines 1142-1289)
- Depends on: Tasks 2 & 3
- Label: **halo-pr-prep**

**Comment by jalengg** [comment]
Task 8 complete: added `tests/integration/test_halo_end_to_end.py` (commit 61424768).

**Tests written:**

Category A — In-Memory (always pass, no external data):
1. `test_full_pipeline_forward` — forward pass returns dict with loss and predictions
2. `test_train_model_runs_one_epoch` — train_model completes one epoch without error
3. `test_checkpoint_saved_after_training` — checkpoint file written after training with val_dataset
4. `test_synthesize_returns_correct_count` — synthesize_dataset(5) returns exactly 5 dicts
5. `test_synthesize_output_structure` — each dict has patient_id and visits (list of lists)
6. `test_pipeline_with_variable_visit_counts` — DataLoader handles patients with 2, 3, 5 visits via _halo_collate_fn
7. `test_model_is_basemodel_instance` — isinstance(model, BaseModel) is True
8. `test_feature_keys` — model.feature_keys == ['visits']

Category B — MIMIC-III (skipped gracefully when data unavailable):
9. `test_mimic3_set_task_returns_nonempty_dataset`
10. `test_mimic3_sample_keys`
11. `test_mimic3_visits_are_nested_lists_of_strings`
12. `test_mimic3_full_pipeline_train_and_synthesize`

**Results on this machine:** 8 passed, 4 skipped (MIMIC-III not available — litdata not installed in test venv)

**Comment by jalengg** [comment]
## What the integration tests are testing — and why they are MECE

The suite in `tests/integration/test_halo_end_to_end.py` is split into two tiers. Here is what each test exercises and the argument for why the set is Mutually Exclusive and Collectively Exhaustive.

---

### Category A — In-Memory (always run, no external data)

These tests use `InMemorySampleDataset` with synthetic code strings so they run in any environment, including CI without MIMIC-III access.

| # | Test | What it asserts |
|---|------|-----------------|
| 1 | `test_full_pipeline_forward` | The data→model→forward path produces a dict with both `"loss"` (finite scalar) and `"predictions"` keys. This is the primary contract: processor output can flow through `_encode_visits` and `HALOModel` without shape errors. |
| 2 | `test_train_model_runs_one_epoch` | The training loop (DataLoader → optimizer → backward → step) completes a full epoch without raising any exception. |
| 3 | `test_checkpoint_saved_after_training` | After training, `os.path.join(save_dir, "halo_model")` exists on disk. Separately verifies the I/O side of training — test 2 only checks the compute side. |
| 4 | `test_synthesize_returns_correct_count` | `synthesize_dataset(num_samples=5)` returns exactly 5 records. Verifies the count contract of the public generation API. |
| 5 | `test_synthesize_output_structure` | Every returned record has `"patient_id"` (str) and `"visits"` (list of list). Verifies the schema contract of the public generation API — test 4 only checks the count. |
| 6 | `test_pipeline_with_variable_visit_counts` | A batch containing patients with 2, 3, and 5 visits respectively flows through `train_model` without a `RuntimeError`. This directly exercises `_halo_collate_fn` (written to fix a real collation crash when patients have unequal visit counts). |
| 7 | `test_model_is_basemodel_instance` | `isinstance(model, BaseModel)` is `True`. Verifies the PyHealth inheritance contract in the integrated execution path, not just at import time. |
| 8 | `test_feature_keys` | `model.feature_keys == ["visits"]`. Verifies that `BaseModel.__init__` correctly derives feature keys from the dataset's `input_schema`, completing the processor→model binding. |

### Category B — MIMIC-III (skipped if data unavailable)

Four tests that repeat the above flow with real patient data: `set_task` produces a non-empty dataset, samples have the expected nested-list structure with at least 2 visits, and a full train+synthesize cycle completes end-to-end.

---

### Why MECE

**Mutually Exclusive** — no two tests assert the same thing:
- Tests 2 and 3 appear similar but are disjoint: 2 checks compute correctness (no crash during training), 3 checks side-effect correctness (file written to disk). A bug that silently skips saving would fail 3 but pass 2.
- Tests 4 and 5 are disjoint: 4 checks count, 5 checks schema. A bug that returns 5 records with the wrong keys would fail 5 but pass 4.
- Test 6 is the only test that exercises variable-length batches. The other training test (2) uses a uniform-visit dataset, so a regression in `_halo_collate_fn` would slip past it.
- Tests 7 and 8 each check a different aspect of the BaseModel integration: 7 checks the inheritance chain, 8 checks the schema-to-processor binding.

**Collectively Exhaustive** — together they cover every integration seam:

```
InMemorySampleDataset
        │
        ▼
NestedSequenceProcessor (vocab, padding)   ← tests 7, 8
        │
        ▼
_halo_collate_fn (variable-length batching) ← test 6
        │
        ▼
HALO.forward / _encode_visits               ← test 1
        │
        ├── train_model (compute)            ← test 2
        │       └── checkpoint (I/O)         ← test 3
        │
        └── synthesize_dataset
                ├── count contract           ← test 4
                └── schema contract          ← test 5
```

The only integration seam not covered by Category A is the real MIMIC-III data path (actual ICD-9 codes, real patient distributions, real vocab sizes) — that is the purpose of Category B.

**Comment by jalengg** [comment]
Closing — Task 8 is complete and verified. Final state: 57 tests passing, 9 skipped (all MIMIC-III data-dependent), 0 failing. Commits: 61424768 (initial), 1b744eb4 (tearDown cleanup + env var path + relative bootstrap paths), d5248de5 (sys.modules stub contamination guard).
